### PR TITLE
Move Dev Environment VPN to Enable Connection to RP and Cluster Private IPs Simultaneously

### PR DESCRIPTION
**Pull request by zacharyljones**
### Which issue this PR addresses:

[USER STORY 14932248](https://dev.azure.com/msazure/AzureRedHatOpenShift/_workitems/edit/14932248/)

### What this PR does / why we need it:

The current dev RP infrastructure has one VPN connected to the dev VNET. This can be used to access the private ip addresses associated with dev clusters, but cannot be used to access private IP addresses associated with the dev RP. Certain actions, like some E2E tests, require access to both simultaneously. Setting the VPN up in a way where the RP and dev clusters can be accessed simultaneously will allow other SREs to test changes more effectively.

### Test plan for issue:

I used the new ARM templates to create a shared RP and then used that RP to...
1. create a private cluster
2. pull the kubeconfig for that cluster and issue kubectl commands against it
3. curl the admin api endpoints to perform various cluster related actions
4. ran e2e tests using the cluster

### Is there any documentation that needs to be updated for this PR?

No - the process for creating this shared RP is exactly the same as before

### CODE REVIEW

**Comment by zacharyljones** [comment]
@coreyperkins @karanmagdani @ross-bryan @nontw - FYSA

**Comment by coreyperkins** [comment]
@zacharyljones - ran local, pulled kubeconfig, and ran e2e tests. This is awesome!

**Review by ross-bryan — COMMENTED** [review]
thank you for this @zacharyljones :+1:

File: pkg/deploy/generator/resources_rp.go
Comment (bennerv) [inline pkg/deploy/generator/resources_rp.go:172]: This logic could be cleaner here.  i.e. default to `Microsoft.Storage` service endpoint, and if production then append the other ones.  don't repeat yourself (DRY)

**Review by bennerv — COMMENTED** [review]
first pass - it's a bit harder to wrap my head around some of these changes as there's a lot of distinguishing factors between production and a local RP.  if possible, an accompanying diagram might make review easier for myself and others.  Nothing too complex, just a bunch of boxes would be nice.

File: pkg/deploy/generator/resources_rp.go
Comment (zacharyljones) [reply pkg/deploy/generator/resources_rp.go]: Updated -  Changing the order that the service endpoints get added did cause similar changes to the production RP template. This shouldn't matter.

Comment (zacharyljones) [reply pkg/deploy/generator/resources_rp.go]: @bennerv I'll begin working on a diagram, and then share a link on slack

Comment (zacharyljones) [reply pkg/deploy/generator/resources_rp.go]: @bennerv Was that diagram I sent over sufficient? Is there anything else you may want or need related to this effort?

**Comment by zacharyljones** [comment]
@karanmagdani @swiencki Is there anything that ya'll would add to this PR based on your testing? Is there anything else that I could help ya'll verify?

**Comment by karanmagdani** [comment]
> @karanmagdani @swiencki Is there anything that ya'll would add to this PR based on your testing? Is there anything else that I could help ya'll verify?

@zacharyljones I tried to set up my own RP with your changes ([v4ss-eastus](https://ms.portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/26c7e39e-2dfa-4854-90f0-6bc88f7a0fb8/resourceGroups/v4ss-eastus/overview)). My 2 attempts at cluster creation created only master nodes and 0 workers nodes. I'm trying to figure out what's causing this at the moment

**Comment by zacharyljones** [comment]
> > @karanmagdani @swiencki Is there anything that ya'll would add to this PR based on your testing? Is there anything else that I could help ya'll verify?
> 
> @zacharyljones I tried to set up my own RP with your changes ([v4ss-eastus](https://ms.portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/26c7e39e-2dfa-4854-90f0-6bc88f7a0fb8/resourceGroups/v4ss-eastus/overview)). My 2 attempts at cluster creation created only master nodes and 0 workers nodes. I'm trying to figure out what's causing this at the moment

I've only seen this when I haven't swapped the SS NSG fast enough. I've been doing this manually, but I believe @swiencki pulled in changes from [PR-2222](https://github.com/Azure/ARO-RP/pull/2222) to prevent having to do this.

**Review by karanmagdani — COMMENTED** [review]
I was able to review these changes by creating my own RP and creating a cluster with PRIVATE_CLUSTER=true in a SimplySecure V2 enabled subscription and run e2e tests locally.

File: pkg/deploy/assets/env-development.json
Comment (karanmagdani) [inline pkg/deploy/assets/env-development.json:154]: We may need to remove the condition  `"[equals(parameters('ciCapacity'), 0)]"`

This block along with [Line 62](https://github.com/Azure/ARO-RP/pull/2238/files#diff-efb7c35dcb2ed4f45e17747bb9d1e87f5e1a99062a63d54c7100a16f16969adcR62) blocks the creation of vpn and public ip address. 

Moving forward with Simply Secure, even in CI deployments we need a VPN.

**Review by karanmagdani — CHANGES_REQUESTED** [review]
Removing condition which blocks creation of VPN in CI environment when running `deploy_env_dev_ci`

File: pkg/deploy/assets/env-development.json
Comment (zacharyljones) [reply pkg/deploy/assets/env-development.json:152]: That makes sense to me - generator and templates have been updated to remove dependency on ci capabilites for VPN

**Comment by zacharyljones** [comment]
@karanmagdani templates and generator have been updated to make the dev vpn vnet peerings dependent on the dev vpn. The gateway related parameters for the vnet peerings require a vpn to exist before creation.

**Comment by zacharyljones** [comment]
@karanmagdani I rebased this PR so it wouldn't conflict with some of @coreyperkins changes that just got merged

**Comment by karanmagdani** [comment]
> @karanmagdani I rebased this PR so it wouldn't conflict with some of @coreyperkins changes that just got merged

Thank you @zacharyljones ! It seems like my PR here https://github.com/CloudFitSoftware/ARO-RP/pull/8/files was not added to this as part of the rebase. Was that expected?

**Comment by zacharyljones** [comment]
> Thank you @zacharyljones ! It seems like my PR here https://github.com/CloudFitSoftware/ARO-RP/pull/8/files was not added to this as part of the rebase. Was that expected?

Ah I missed your PR. I'll take care of it.

File: pkg/deploy/assets/env-development.json
Comment (karanmagdani) [inline pkg/deploy/assets/env-development.json:105]: dev-vpn-vnet is a newly added vnet which houses the VirtualNetworkGateway (VPN) client. Please refer to this diagram for how vnets are setup moving forward https://lucid.app/lucidchart/34183535-857f-45df-b68d-45772d0a27e3/edit?view_items=8caD-VenYEWG&invitationId=inv_282dd721-1a88-45af-8eba-f9c3938d3235#

File: pkg/deploy/generator/resources.go
Comment (karanmagdani) [inline pkg/deploy/generator/resources.go:116]: [Gateway transit](https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-peering-overview#gateways-and-on-premises-connectivity) is a peering property that lets one virtual network use the VPN gateway in the peered virtual network for cross-premises or VNet-to-VNet connectivity

We want to be able to Allow Gateway Transit / Use Remote Gateways for the particular peerings: 

```
dev-vpn-vnet -> dev-vnet [Allow Gateway Transit = True, Use Remote Gateways = False]
dev-vnet -> dev-vpn-vnet [Allow Gateway Transit = False, Use Remote Gateways = True]
dev-vpn-vnet -> rp-vnet [Allow Gateway Transit = True, Use Remote Gateways = False]
rp-vnet -> dev-vpn-vnet [Allow Gateway Transit = False, Use Remote Gateways = True]
```

**Review by karanmagdani — COMMENTED** [review]
Since this PR has quite a few changes. I want to leave this review as a comment so approvers can get context on the changes and hopefully simplify the review process. 

All these changes affect development mode and CI-E2E only. Should not have any impact on Production clusters. 

- Once this is reviewed and merged, our CI E2E RP infra would be recreated as part of this WI: https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/15093590/
- At some point, we should also consider recreating our Dev RP infra so we have the same setup across different environments. 

CC @ross-bryan @bennerv Would appreciate a reviews on this one

**Review by karanmagdani — DISMISSED** [review]

File: pkg/deploy/generator/resources_dev.go
Comment (ross-bryan) [inline pkg/deploy/generator/resources_dev.go]: /16

File: pkg/deploy/generator/resources.go
Comment (zacharyljones) [reply pkg/deploy/generator/resources.go:116]: @karanmagdani somehow the dev-vnet still had a /9 in the generator but all the templates were correct. I changed the /9 to a /16 to match what we had talked about.

Comment (karanmagdani) [reply pkg/deploy/generator/resources.go:116]: Thanks @zacharyljones , i think i may have missed that as part of my PR against your branch. Also feel free to add any comments  that you believe will help reviewers

**Review by ross-bryan — APPROVED** [review]
lgtm

**Review by karanmagdani — APPROVED** [review]
LGTM! Thanks a lot @zacharyljones

**Pull request merged**
